utils/fork-tree/src/lib.rs

```diff
@@ -289,6 +289,10 @@ where
 	/// index in the traverse path goes last. If a node is found that matches the predicate
 	/// the returned path should always contain at least one index, otherwise `None` is
 	/// returned.
+	// WARNING: some users of this method (i.e. Babe epoch changes tree) currently silently
+	// rely on a **post-order DFS** traversal of the tree. In particular via the
+	// `is_descendent_of` that when used after a warp-sync may query the backend for a
+	// root that is not present.
 	pub fn find_node_index_where<F, E, P>(
```

Comment:
Maybe the traversal could be chosen by the user? 

Then we don't need all these warnings etc ;)

Reply:
That is a possibility, but IMO it has two major drawbacks:
1. For (design) coherence all the tree methods that perform a traversal (almost all) should introduce this parameter to allow the user to choose. Otherwise the user may wonder why we reserved this special treatment `find_node_index`.
2. Allowing the user to choose between, let's say two traversal methods, implies that we should implement both traversal methods for all the methods exposing it. And then internally match and jump to the chosen one.

In the end, I think that a proper solution should not expose the traversal method. In other words is the user assumption that is not correct, the method should work regardless of the traversal method used (i.e. it had to work without this fix, that is more close to a hack to fulfill the caller expectations).

An idea could be to:
1. apply this fairly "dirty" fix
2. immediately open a follow up Issue where we solve the problem properly, i.e. the user should not "assume" (see below)